tools/floodgate/js/promote.js

```diff
@@ -0,0 +1,169 @@
+import {
+  createFolder,
+  getAuthorizedRequestOption,
+  saveFile,
+  updateExcelTable,
+  validateConnection,
+} from '../../loc/sharepoint.js';
+import {
+  hideButtons,
+  loadingON,
+  showButtons,
+  simulatePreview,
+} from '../../loc/utils.js';
+import { getConfig as getFloodgateConfig } from './config.js';
+import { ACTION_BUTTON_IDS } from './ui.js';
+import { getFile, handleExtension } from './utils.js';
+
+/**
+ * Copies the Floodgated files back to the main content tree.
+ * Creates intermediate folders if needed.
+ */
+async function promoteCopy(srcPath, destinationFolder) {
+  validateConnection();
+  await createFolder(destinationFolder);
+  const { sp } = await getFloodgateConfig();
+  const destRootFolder = `${sp.api.file.copy.baseURI}`.split('/').pop();
+
+  const payload = { ...sp.api.file.copy.payload, parentReference: { path: `${destRootFolder}${destinationFolder}` } };
+  const options = getAuthorizedRequestOption({
+    method: sp.api.file.copy.method,
+    body: JSON.stringify(payload),
+  });
+
+  // copy source is the pink directory for promote
+  const copyStatusInfo = await fetch(`${sp.api.file.copy.fgBaseURI}${srcPath}:/copy`, options);
+  const statusUrl = copyStatusInfo.headers.get('Location');
+  let copySuccess = false;
+  let copyStatusJson = {};
+  while (statusUrl && !copySuccess && copyStatusJson.status !== 'failed') {
+    // eslint-disable-next-line no-await-in-loop
+    const status = await fetch(statusUrl);
+    if (status.ok) {
+      // eslint-disable-next-line no-await-in-loop
+      copyStatusJson = await status.json();
+      copySuccess = copyStatusJson.status === 'completed';
+    }
+  }
+  return copySuccess;
+}
+
+/**
+ * Iteratively finds all files under a specified root folder.
+ */
+async function findAllFloodgatedFiles(baseURI, options, rootFolder, fgFiles, fgFolders) {
+  while (fgFolders.length !== 0) {
+    const uri = `${baseURI}${fgFolders.shift()}:/children`;
+    // eslint-disable-next-line no-await-in-loop
+    const res = await fetch(uri, options);
+    if (res.ok) {
+      // eslint-disable-next-line no-await-in-loop
+      const json = await res.json();
+      const driveItems = json.value;
+      if (driveItems) {
+        driveItems.forEach((item) => {
+          const itemPath = `${item.parentReference.path.replace(`/drive/root:/${rootFolder}`, '')}/${item.name}`;
+          if (item.folder) {
+            // it is a folder
+            fgFolders.push(itemPath);
+          } else {
+            const downloadUrl = item['@microsoft.graph.downloadUrl'];
+            fgFiles.push({ fileDownloadUrl: downloadUrl, filePath: itemPath });
+          }
+        });
+      }
```

Comment:
Nit: optional chaining
```suggestion
    driveItems?.forEach((item) => {
      const itemPath = `${item.parentReference.path.replace(`/drive/root:/${rootFolder}`, '')}/${item.name}`;
      if (item.folder) {
        // it is a folder
        fgFolders.push(itemPath);
      } else {
        const downloadUrl = item['@microsoft.graph.downloadUrl'];
        fgFiles.push({ fileDownloadUrl: downloadUrl, filePath: itemPath });
      }
    });
```